File: src/packages/control-ui/src/api/leak_guard_supervisor.ts
export interface UiLeakRecord {
  type: string;
  dstIp: string;
  iface: string;
}

export class LeakGuardSupervisor {
  public killswitchActive: boolean = true;
  private leaks: UiLeakRecord[] = [];
  public tunnelIface: string;
  public allowedDns: string[];
  public blockIpv6: boolean;
  constructor(tunnelIface: string, allowedDns: string[], blockIpv6: boolean = true) {
    this.tunnelIface = tunnelIface;
    this.allowedDns = allowedDns;
    this.blockIpv6 = blockIpv6;
  }

  validateOutbound(dstIp: string, dstPort: number, iface: string): boolean {
    if (!this.killswitchActive) return true;

    if (iface !== this.tunnelIface) {
      if ([53, 853, 5353].includes(dstPort)) {
        if (!this.allowedDns.includes(dstIp)) {
          this.leaks.push({ type: 'DNS_LEAK', dstIp, iface });
          return false;
        }
      }

      if (this.blockIpv6 && dstIp.includes(':')) {
        this.leaks.push({ type: 'IPV6_LEAK', dstIp, iface });
        return false;
      }
    }

    return true;
  }

  totalLeaks(): number {
    return this.leaks.length;
  }
}

File: labs/mobile/donor-archive/tunnel/LeakGuardSupervisor.kt
package com.luminet.android.tunnel

data class AndroidLeakReport(
    val leakType: String,
    val targetIp: String,
    val iface: String
)

class LeakGuardSupervisor(
    private val tunnelIface: String,
    private val allowedDns: List<String>,
    private val blockIpv6: Boolean = true
) {
    var killswitchActive: Boolean = true
    private val leaks = mutableListOf<AndroidLeakReport>()

    fun validateOutbound(dstIp: String, dstPort: Int, iface: String): Boolean {
        if (!killswitchActive) return true

        if (iface != tunnelIface) {
            // DNS leak check
            if (dstPort in listOf(53, 853, 5353)) {
                if (!allowedDns.contains(dstIp)) {
                    leaks.add(AndroidLeakReport("DNS_LEAK", dstIp, iface))
                    return false
                }
            }

            // IPv6 leak check
            if (blockIpv6 && dstIp.contains(":")) {
                leaks.add(AndroidLeakReport("IPV6_LEAK", dstIp, iface))
                return false
            }
        }

        return true
    }

    fun totalLeaks(): Int = leaks.size
}
